core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1531,6 +1531,8 @@ class Partition(val topicPartition: TopicPartition,
             isrState = CommittedIsr(leaderAndIsr.isr.toSet)
             zkVersion = leaderAndIsr.zkVersion
             info(s"ISR updated to ${isrState.isr.mkString(",")} and version updated to [$zkVersion]")
+            leaderLogIfLocal.exists(leaderLog => maybeIncrementLeaderHW(leaderLog))
```

Comment:
You may also need to call tryCompleteDelayedRequests if HW is incremented. Is deplayed request checked periodically already? I know for other HW updates, it is normally followed with tryCompleteDelayedRequests

Reply:
Good point! tryCompleteDelayedRequests will be called on each fetch request. Since fetch is a continuously running process, I don't think we need to explicitly call it here.

Reply:
"a continuously running process" -- do you mean even the fetcher has fetched no new records, it still calls completeDelayedRequests for all delayed requests? And in some corner cases, the fetcher may terminate? If no other concerns, I feel there is no harm to call tryCompleteDelayedRequests immediately after shrinking ISR.

Reply:
Adding completeDelayedRequests will introduce the deadlock.
Check this one: https://issues.apache.org/jira/browse/KAFKA-13254

This is because there is one step in completeDelayedFetch that acquires LeaderAndIsr lock, and the handleAlterIsrResponse also holds the same lock.

Reply:
> do you mean even the fetcher has fetched no new records, it still calls completeDelayedRequests for all delayed requests?

This one, however, is a good question. I rechecked the code and believe completeDelayedRequests shouldn't be called if there is no new data. In that case, I'm not sure why incrementHW will complete those delayed requests. Another mystery...

I'm considering pulling in upstream changes (if I'm not able to figure out the above mystery), as those fixes makes more sense.